133 take 2@filter panel refactor@main

asbates opened the pull request:
Adds a numeric range input to range filter state cards so range can be set manually. This is a redo of #226


Fixes #133

chlebowa commented on R/FilterStateRange.R:
```diff
@@ -493,6 +526,74 @@ RangeFilterState <- R6::R6Class( # nolint
+                )
+                return(NULL)
+              }
+              # all.equal not enough here b/c tolerance
+              # all.equal(0.000000001, 0) is TRUE
+              out_of_range <- isFALSE(identical(
```
`all.equal` was there exactly to allow the tolerance

chlebowa replied:
Also, with `identical` `isFALSE` is redundant but with `all.equal` `isFALSE` will fail and `!isTRUE` is necessary.

chlebowa reviewed (changes requested):
Invalid inputs cannot crash the whole app. There has to be some way to block that error. `shinyvalidate` is one option but I don't know whether we want to add it to `teal.slice`.

gogonzo commented:
> Invalid inputs cannot crash the whole app. There has to be some way to block that error. shinyvalidate is one option but I don't know whether we want to add it to teal.slice.

We can't use `shinyvalidate` as FilterPanel must not return any shiny-validate errors - this would have a serious consequences downstream in `teal` and `teal_module`. So, if user sets the wrong value it should be automatically set to the latest selected with notification

gogonzo commented on R/FilterStateRange.R:
```diff
@@ -493,6 +526,74 @@ RangeFilterState <- R6::R6Class( # nolint
+                shinyWidgets::updateNumericRangeInput(
+                  session = session,
+                  inputId = "selection_manual",
+                  value = self$get_selected()
```
this is why it fails
```suggestion
                  value = private$get_selected()
```

gogonzo replied:
please replace all `self$get_selected` with `private$get_selected` and then module should be robust enough

chlebowa reviewed (changes requested):
Numeric selection states on app start do not follow api specification:

```
default_filters <- filter_settings(
  filter_var("iris", "Sepal.Length", sel = c(5.1, 6.4), fixed = FALSE),
  filter_var("iris", "Sepal.Width", sel = c(2.5, 4.0), fixed = FALSE, disabled = TRUE),
  filter_var("iris", "Petal.Length", sel = c(4.5, 5.1), fixed = TRUE),
  filter_var("iris", "Petal.Width", sel = c(0.3, 1.8), fixed = TRUE, disabled = TRUE),
  filter_var("iris", "Species", sel = c("versicolor", "virginica"), fixed = FALSE),
  filter_var("iris", "Species2", sel = c("versicolor", "virginica"), fixed = TRUE)
)
```

![image](https://github.com/insightsengineering/teal.slice/assets/114988527/f03e46c3-b18a-4863-91ff-23536c859155)

Also, `choices` become limited.

chlebowa commented:
I think `RangeFilterState$set_choices` is broken but that's a different issue. I will look into it a little more.

chlebowa commented:
> Numeric selection states on app start do not follow api specification:
(...)

My mistake, the numbers in the app are the actual data ranges.

asbates commented:
> Numeric selection states on app start do not follow api specification:
> 
> ```
> default_filters <- filter_settings(
>   filter_var("iris", "Sepal.Length", sel = c(5.1, 6.4), fixed = FALSE),
>   filter_var("iris", "Sepal.Width", sel = c(2.5, 4.0), fixed = FALSE, disabled = TRUE),
>   filter_var("iris", "Petal.Length", sel = c(4.5, 5.1), fixed = TRUE),
>   filter_var("iris", "Petal.Width", sel = c(0.3, 1.8), fixed = TRUE, disabled = TRUE),
>   filter_var("iris", "Species", sel = c("versicolor", "virginica"), fixed = FALSE),
>   filter_var("iris", "Species2", sel = c("versicolor", "virginica"), fixed = TRUE)
> )
> ```
> 
> ![image](https://user-images.githubusercontent.com/114988527/237153716-f03e46c3-b18a-4863-91ff-23536c859155.png)
> 
> Also, `choices` become limited.

Huh, it's working for me. This is what I get on launch.

![image](https://github.com/insightsengineering/teal.slice/assets/36482959/7fab76c6-7188-4cae-9554-c3bcf8e30ef7)

And after enabling

![image](https://github.com/insightsengineering/teal.slice/assets/36482959/8ed80e26-8060-47ff-94c5-2d272c705859)

chlebowa commented:
> Huh, it's working for me. This is what I get on launch.
> 
> ![image](https://user-images.githubusercontent.com/36482959/237158006-7fab76c6-7188-4cae-9554-c3bcf8e30ef7.png)

uh-oh

donyunardi commented:
I have the same result with @asbates.
This is what I see upon launch:

<img src="https://github.com/insightsengineering/teal.slice/assets/8597300/fb70e58b-7a88-4538-9c37-3d74be39f899" width="50%" height="50%" />

chlebowa commented:
All right, my session was littered and another object got pulled into the app. Works fine for me as well.
